src/plus/gk/serverConnection.ts

```diff
@@ -169,28 +192,180 @@ export class ServerConnection implements Disposable {
 					url = `${url}?${options.query}`;
 				}
 			}
-			// TODO@eamodio handle common response errors
 
-			return this.fetch(
+			const rsp = await this.fetch(
 				url,
 				{
 					...init,
 					headers: headers as HeadersInit,
 				},
 				options,
 			);
+			if (!rsp.ok) {
+				await this.handleUnsuccessfulResponse(rsp, scope);
+			} else {
+				this.resetRequestExceptionCount();
+			}
+			return rsp;
 		} catch (ex) {
-			Logger.error(ex, scope);
+			this.handleRequestError('gitkraken', ex, scope);
 			throw ex;
 		}
 	}
 
+	private buildProviderRequestRateLimitError(token: string | undefined, ex: RequestError) {
+		let resetAt: number | undefined;
+
+		const reset = ex.response?.headers?.['x-ratelimit-reset'];
+		if (reset != null) {
+			resetAt = parseInt(reset, 10);
+			if (Number.isNaN(resetAt)) {
+				resetAt = undefined;
+			}
+		}
+		return new ProviderRequestRateLimitError(ex, token, resetAt);
+	}
+
+	private async handleUnsuccessfulResponse(rsp: Response, scope: LogScope | undefined): Promise<void> {
+		// Too Many Requests
+		if (rsp.status == 429) {
+			this.trackRequestException();
+			return;
+		}
+
+		// Forbidden
+		if (rsp.status == 403) {
+			if (rsp.statusText.includes('rate limit')) {
+				this.trackRequestException();
+			}
+			return;
+		}
+
+		// Internal Server Error
+		if (rsp.status == 500) {
+			this.trackRequestException();
+			void showIntegrationRequestFailed500WarningMessage(
+				'GitKraken failed to respond and might be experiencing issues. Please visit the [GitKraken status page](https://cloud.gitkrakenstatus.com) for more information.',
+			);
+			return;
+		}
+
+		const content = await rsp.text();
+
+		// Bad Gateway
+		if (rsp.status == 502) {
+			Logger.error(`GitKraken request failed: ${content} (${rsp.statusText})`, scope);
+			if (content.includes('timeout')) {
+				this.trackRequestException();
+				void showIntegrationRequestTimedOutWarningMessage('GitKraken');
+			}
+			return;
+		}
+
+		// Service Unavailable
+		if (rsp.status == 503) {
+			Logger.error(`GitKraken request failed: ${content} (${rsp.statusText})`, scope);
+			this.trackRequestException();
+			void showIntegrationRequestFailed500WarningMessage(
+				'GitKraken failed to respond and might be experiencing issues. Please visit the [GitKraken status page](https://cloud.gitkrakenstatus.com) for more information.',
+			);
+			return;
+		}
+
+		if (rsp.status >= 400 && rsp.status < 500) {
+			return;
+		}
+
+		if (Logger.isDebugging) {
+			void window.showErrorMessage(`GitKraken request failed: ${content} (${rsp.statusText})`);
+		}
+	}
+
+	private handleRequestError(
```

Comment:
Since this is currently for GK calls only (as the messaging and tracking indicate), we should name this something more specific, like `handleGkRequestError`, so that it is not accidentally applied in the future to fetch calls outside of the GK infrastructure.

Note: Same advice would apply for `handleUnsuccessfulResponse`.

Reply:
done